### PR TITLE
Update STYLE-GUIDE.md

### PR DESCRIPTION
**Pull request by wora**
Mention the CRD kind must match proto message.

### CODE REVIEW

File: STYLE-GUIDE.md [hunk 1]
```diff
@@ -38,7 +38,8 @@ naming convention:
 ## Versioning
 
 When defining Kubernetes Custom Resource Definition (CRD) using
-proto3, the proto versioning should match the Kubernetes versioning,
+proto3, the proto versioning must match the Kubernetes versioning,
+and the proto message type must match the CRD kind name.
```
Comment (ayj) [inline STYLE-GUIDE.md]: Should group names and version also be consistent with CRDs? We've defined versioned package in istio/api but these all map to the same CRD group and version (config.istio.io). 

For example, does the following mapping make sense?

```golang
group   = routing.api.istio.io
version = v1alpha2
kind    = RouteRule

group   = rbac.api.istio.io
version = v1alpha1
kind    = ServiceRoleBinding
```

Comment (xiaolanz) [reply STYLE-GUIDE.md]: @ayj  this is exactly what I am proposing: https://groups.google.com/d/msg/istio-config/CzH6nRBT_o0/3lvIBao_AgAJ

I like this to be consistent with api group too.

Comment (wora) [reply STYLE-GUIDE.md]: Yes. I should have made that more clear. The CRD apiVersion must match proto package name, minus the `.io` suffix, and the ordering. I will update the text.

Comment (xiaolanz) [reply STYLE-GUIDE.md]: These are apiGroup. not apiVersion.

File: STYLE-GUIDE.md [hunk 2]
```diff
@@ -38,8 +38,13 @@ naming convention:
 ## Versioning
 
 When defining Kubernetes Custom Resource Definition (CRD) using
-proto3, the proto versioning should match the Kubernetes versioning,
-see the following example.
+`proto3`, follow the following guidelines:
+* The proto `package` name must match the Kubernetes `apiVersion`,
+  excluding the `.io` DNS suffix and reversing the DNS segment
+  ordering.
+* The proto message type must match the CRD `kind` name.
```
Comment (ayj) [inline STYLE-GUIDE.md]: I'm not sure if its worth highlighting here, but the `apiVersion` as shown in the YAML snippet is the  concatenation of the Kubernetes API _group_ and _version_. In the example below, `config.istio.io` is the group and `v1alpha` is the version of that group. Together they identify the versioned schema the object should have.

**Comment by wora** [comment]
I addressed the concern of apiVersion vs group/version.

File: STYLE-GUIDE.md (hunk 2 above)
Comment (xiaolanz) [reply STYLE-GUIDE.md]: Maybe in the yaml example, listing out a full CRD definition. We probably also need to make rule that CRD resource name should be in format of plurals + api group. 

apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: serviceclasses.config.istio.io
spec:
  group: config.istio.io
  version: v1alpha1
  names:
    plural: serviceclasses
    singular: serviceclass
    kind: ServiceClass

Comment (wora) [reply STYLE-GUIDE.md]: Can we address this question in another PR? Let's make some incremental improvement first.

**Comment by wora** [comment]
@geeknoid Can you merge this PR?

**Comment by ozevren** [comment]
With this guidance, is it possible to create meta-APIs such as LROs? It looks like requiring 1-1 mapping between proto package name and apigroup seems to be work against this.

File: STYLE-GUIDE.md [hunk 3]
```diff
@@ -38,8 +38,14 @@ naming convention:
 ## Versioning
 
 When defining Kubernetes Custom Resource Definition (CRD) using
-proto3, the proto versioning should match the Kubernetes versioning,
-see the following example.
+`proto3`, follow the following guidelines:
```
Comment (geeknoid) [inline STYLE-GUIDE.md]: , follow these guidelines:

Comment (geeknoid) [inline STYLE-GUIDE.md]: , follow these guidelines:

**Comment by geeknoid** [comment]
@ozevren Doesn't the sentence "When defining Kubernetes Custom Resource Definition" cover the case you're worried about? LRO is specifically not a CRD.

**Comment by ozevren** [comment]
Meant to use LRO as an example. It is a meta-API. We're using CRDs to
create a public API surface (which happens to be backed by the API Server).
Adding this constraint implies you cannot create a meta set of protos to
use as top-level mix-in resources. And, it also means you cannot re-use
top-level CRD protos for another "API group" without violating this rule
(i.e. think of the Config Click Stop-0 we've been talking about).

**Comment by wora** [comment]
K8s APIs support watch semantics, which is equivalent to LRO. I don't think watch semantics is implemented by any api group. It is more like a design pattern. We need to address it next, but I don't think it is blocked by this discussion.

**Comment by geeknoid** [comment]
Oz was using LRO as an example of a pattern, not as a specific case.

As stated right now, the guide doesn't allow for common definitions that
can be shared between multiple APIs because we are forcing a 1:1 mapping
pattern. We just need to add more nuance to the text to allow for those
cases.

**Comment by rshriram** [comment]
I have no idea about LROs, but as such, I am not against this proposal for routing section.

**Comment by wora** [comment]
Can someone merge this PR or let me any further feedback? Thanks.

**Pull request merged**